scalacheck/src/main/scala/karat/scalacheck/Scalacheck.scala

```diff
@@ -28,17 +30,47 @@ object Scalacheck {
       test.invoke(value)
   }
 
-  def checkFormula[Action, State, Response](actions: List[Action], initial: State, step: (Action, State) => Step[State, Response])(
+  def checkFormula[Action, State, Response](actions: List[Action], initial: State, step: (Action, State) => Option[Step[State, Response]])(
     formula: Formula[Info[Action, State, Response]]
-  ): Prop = {
+  ): Prop.Result = {
     val problem = CheckKt.check[Action, State, Response, Prop.Result, Prop.Result](
```

Comment:
What do you think about the next action creating an intermediate layer that doesn't need to deal with nulls or unsafe castings?

Reply:
Sure! Let's try to make the use as nice as possible from Scala.